extensions/big-data-cluster/src/bigDataCluster/dialog/addControllerDialog.ts

```diff
@@ -0,0 +1,500 @@
+		if (response && response.request) {
+			let masterInstance: IEndPoint = undefined;
+			if (response.endPoints) {
+				masterInstance = response.endPoints.find(e => e.name && e.name === 'sql-server-master');
+			}
+			this.treeDataProvider.addController(response.request.url, response.request.username,
+				response.request.password, rememberPassword, masterInstance);
+			await this.treeDataProvider.saveControllers();
+		}
+	}
+
+	public async onError(error: IControllerError): Promise<void> {
+		// implement
+	}
+
+	public async onCancel(): Promise<void> {
+		if (this.node) {
+			this.node.refresh();
+		}
+	}
+}
+
+export class AddControllerDialog {
+
+	private readonly DialogTitle: string = localize('textAddNewController', 'Add New Controller');
+	private readonly SignInButtonText: string = localize('textAdd', 'Add');
+	private readonly CancelButtonText: string = localize('textCancel', 'Cancel');
+
+	private dialog: azdata.window.Dialog;
+	private uiModelBuilder: azdata.ModelBuilder;
+
+	private urlInputBox: azdata.InputBoxComponent;
+	private usernameInputBox: azdata.InputBoxComponent;
+	private passwordInputBox: azdata.InputBoxComponent;
+	private rememberPwCheckBox: azdata.CheckBoxComponent;
+	private errorTextContainer: azdata.FlexContainer;
+
+	constructor(private model: AddControllerDialogModel) {
+	}
+
+	public showDialog(): void {
+		this.createDialog();
+		this.toggleSignInButton();
+		this.toggleErrorText();
+		azdata.window.openDialog(this.dialog);
+	}
+
+	private createDialog(): void {
+		this.dialog = azdata.window.createModelViewDialog(this.DialogTitle);
+		this.dialog.registerContent(async view => {
+			this.uiModelBuilder = view.modelBuilder;
+
+			let titledContainer = new TitledContainer(this.uiModelBuilder);
+			titledContainer.title = localize('textSignInToController', 'Sign In to Controller');
+			titledContainer.setTitleMargin(6, 0, 11, 0);
+			titledContainer.setPadding(15, 30, 0, 30);
+			titledContainer.contentTopLine = true;
+
+			let wrapperContainer: azdata.FlexContainer = this.uiModelBuilder.flexContainer()
+				.withLayout(<azdata.FlexLayout>{
+					flexFlow: 'column',
+					alignItems: 'stretch',
+					width: '100%'
+				})
+				.component();
+
+			let urlInputContainer: azdata.FlexContainer = this.uiModelBuilder.flexContainer()
+				.withLayout(<azdata.FlexLayout>{
+					flexFlow: 'row',
+					alignItems: 'center',
+					width: '100%'
+				})
+				.component();
+
+			let urlInputText = this.uiModelBuilder.text()
+				.withProperties<azdata.TextComponentProperties>({ value: 'URL' }).component();
+
+			this.urlInputBox = this.uiModelBuilder.inputBox()
+				.withProperties<azdata.InputBoxProperties>({ placeHolder: 'url', value: this.model.prefilledUrl }).component();
+			this.urlInputBox.onTextChanged(e => this.toggleSignInButton());
+
+			urlInputContainer.addItem(urlInputText, {
+				flex: '1, 1, 0%',
+				CSSStyles: {
+					'justify-content': 'flex-start',
```

Comment:
Aren't justify-content and align-items properties all set on the parent container? Shouldn't you just set `justifyCOntent` on the urlInputContainer and skip setting here?

Reply:
removed.